### PR TITLE
Support query words limit

### PR DESCRIPTION
**Pull request by shekhirin**
Resolves https://github.com/meilisearch/milli/issues/147

### CODE REVIEW

**Comment by Kerollmops** [comment]
Hey @shekhirin, are you sure that you need https://github.com/meilisearch/Tokenizer/pull/43 to be able to implement this feature by following @LegendreM's [comment](https://github.com/meilisearch/milli/issues/147#issuecomment-820289885)?

**Comment by shekhirin** [comment]
@Kerollmops yeah, no need to change Tokenizer for that kind of solving the problem. I made an implementation of that idea in the latest commit.

Still not sure about constant limit of 10 words. Shouldn't be better to pass it from somewhere higher or even make it configurable?

File: milli/src/search/query_tree.rs [hunk 1]
```diff
@@ -476,13 +484,15 @@ impl PrimitiveQueryPart {
...
+fn create_primitive_query(query: TokenStream, stop_words: Option<Set<&[u8]>>, words_limit: Option<usize>) -> PrimitiveQuery {
     let mut primitive_query = Vec::new();
     let mut phrase = Vec::new();
     let mut quoted = false;
 
+    let parts_limit = words_limit.unwrap_or(usize::MAX);
+
     let mut peekable = query.peekable();
-    while let Some(token) = peekable.next() {
+    while let (Some(token), true) = (peekable.next(), primitive_query.len() < parts_limit) {
```
Comment (LegendreM) [inline milli/src/search/query_tree.rs:495]:
```suggestion
    while let Some(token) = peekable.next() {
        // early return if word limit is exceded
        if primitive_query.len() >= parts_limit { return primitive_query }
```
I find it clearer and allow you to rollback Line 526 🙂

File: milli/src/search/query_tree.rs [hunk 2]
```diff
@@ -513,7 +523,7 @@ fn create_primitive_query(query: TokenStream, stop_words: Option<Set<&[u8]>>) ->
     }
 
     // If a quote is never closed, we consider all of the end of the query as a phrase.
-    if !phrase.is_empty() {
+    if !phrase.is_empty() && primitive_query.len() < parts_limit {
```
Comment (LegendreM) [inline milli/src/search/query_tree.rs:526]:
```suggestion
    if !phrase.is_empty() {
```

File: milli/src/search/mod.rs [hunk 3]
```diff
@@ -241,7 +254,7 @@ pub fn word_derivations<'c>(
             }
 
             Ok(entry.insert(derived_words))
-        },
+        }
```
Comment (LegendreM) [inline milli/src/search/mod.rs:257]:
```suggestion
        },
```
We chose to keep the coma on every match cases

File: milli/src/search/mod.rs [hunk 4]
```diff
@@ -182,7 +190,11 @@ impl<'a> Search<'a> {
             excluded_documents = candidates.into_excluded();
         }
 
-        Ok(SearchResult { matching_words, candidates: initial_candidates, documents_ids })
+        Ok(SearchResult {
+            matching_words,
+            candidates: initial_candidates,
+            documents_ids,
+        })
```
Comment (LegendreM) [inline milli/src/search/mod.rs:197]:
```suggestion
        Ok(SearchResult { matching_words, candidates: initial_candidates, documents_ids })
```

File: milli/src/search/mod.rs [hunk 5]
```diff
@@ -105,7 +115,7 @@ impl<'a> Search<'a> {
                 let result = analyzer.analyze(query);
                 let tokens = result.tokens();
                 builder.build(tokens)?
-            },
+            }
```
Comment (LegendreM) [inline milli/src/search/mod.rs:118]:
```suggestion
            },
```
We chose to keep the coma on every match cases

**Review by LegendreM — CHANGES_REQUESTED** [review]

**Review by LegendreM — APPROVED** [review]
Good for me!

Thank you @shekhirin for your contribution!

**Comment by Kerollmops** [comment]
Thank you again @shekhirin 🎉🌮

**Pull request merged**